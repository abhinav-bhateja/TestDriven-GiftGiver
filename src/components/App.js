import React, {Component} from "react";
import {Button} from "react-bootstrap";
import {max_number} from '../helper/index'
import Gift from './Gift'

class App extends Component {
    constructor() {
        super();
        this.state = {gifts: []};
    }

    addGift = () => {
        const {gifts} = this.state;
        gifts.push({
            id: max_number(this.state.gifts.map(gift => gift.id)) + 1
        })
        ;
        this.setState({gifts})
    }

    removeGift = id => {
        const gifts = this.state.gifts.filter(gift => gift.id !== id);
        this.setState({gifts})
    }

    render() {
        return (
            <div>

                <div style={{
                    marginTop: "2rem",
                    marginBottom: "2rem",
                    display: "flex",
                    width: "100vw",
                    alignItems: "center",
                    justifyContent: "center"
                }}>
                    <h2>Gift Giver</h2>
                </div>

                <div className="gift-list">
                    {
                        this.state.gifts.map(gift => {
                            return (
                                <Gift
                                    key={gift.id}
                                    gift={gift}
                                    removeGift={this.removeGift}
                                />
                            )
                        })
                    }
                </div>

                <div style={{display: "flex", width: "100vw", alignItems: "center", justifyContent: "center"}}>
                    <Button className="btn-add" onClick={this.addGift}>Add Gift</Button>
                </div>

            </div>
        )
    }
}

export default App;