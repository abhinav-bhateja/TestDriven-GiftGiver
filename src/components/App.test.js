import React from "react";
import {shallow} from "enzyme";
import App from './App'

describe('App', () => {
    const app = shallow(<App/>)

    it('renders correctly', () => {
        expect(app).toMatchSnapshot();
    });

    it('should initialize the `state` with an empty list of gifts', function () {
        expect(app.state().gifts).toEqual([])
    });

    describe('when clicking the `add-gift` button', () => {
        const id = 1;
        beforeEach(() => {
            app.find('.btn-add').simulate('click');
        })

        afterEach(() => {
            app.setState({gifts: []})
        })

        it('should add new gift to `state` when clicking the `add gift` button', function () {
            expect(app.state().gifts).toEqual([{id}])
        });

        it('should adds a new gifts to the rendered list when clicking the `add gift` button', function () {
            expect(app.find('.gift-list').children().length).toEqual(id);
        });

        it('should create a gift component', function () {
            expect(app.find('Gift').exists()).toBe(true);
        });

        describe('when the user wants to remove the added gift', () => {
            beforeEach(() => {
                app.instance().removeGift(id);
            })

            it('should remove the gift from `state`', function () {
                expect(app.state().gifts).toEqual([]);
            });
        })
    });


})
