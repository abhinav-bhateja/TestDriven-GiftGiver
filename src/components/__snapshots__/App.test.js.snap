// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "center",
        "marginBottom": "2rem",
        "marginTop": "2rem",
        "width": "100vw",
      }
    }
  >
    <h2>
      Gift Giver
    </h2>
  </div>
  <div
    className="gift-list"
  />
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "center",
        "width": "100vw",
      }
    }
  >
    <Button
      active={false}
      className="btn-add"
      disabled={false}
      onClick={[Function]}
      variant="primary"
    >
      Add Gift
    </Button>
  </div>
</div>
`;
